h2/src/main/org/h2/mvstore/db/LobStorageMap.java

```diff
@@ -68,83 +71,96 @@
 
     private StreamStore streamStore;
 
+
     public LobStorageMap(Database database) {
         this.database = database;
-    }
-
-    @Override
-    public void init() {
-        if (init) {
-            return;
-        }
-        init = true;
         Store s = database.getStore();
-        MVStore mvStore;
         if (s == null) {
             // in-memory database
             mvStore = MVStore.open(null);
         } else {
             mvStore = s.getMvStore();
         }
-        lobMap = mvStore.openMap("lobMap");
-        refMap = mvStore.openMap("lobRef");
+    }
 
-        /* The stream store data map.
-         *
-         * Key: stream store block id (long).
-         * Value: data (byte[]).
-         */
-        MVMap<Long, byte[]> dataMap = mvStore.openMap("lobData");
-        streamStore = new StreamStore(dataMap);
-        // garbage collection of the last blocks
-        if (database.isReadOnly()) {
-            return;
-        }
-        if (dataMap.isEmpty()) {
-            return;
-        }
-        // search for the last block
-        // (in theory, only the latest lob can have unreferenced blocks,
-        // but the latest lob could be a copy of another one, and
-        // we don't know that, so we iterate over all lobs)
-        long lastUsedKey = -1;
-        for (Entry<Long, Object[]> e : lobMap.entrySet()) {
-            long lobId = e.getKey();
-            Object[] v = e.getValue();
-            byte[] id = (byte[]) v[0];
-            long max = streamStore.getMaxBlockKey(id);
-            // a lob may not have a referenced blocks if data is kept inline
-            if (max != -1 && max > lastUsedKey) {
-                lastUsedKey = max;
+    @Override
+    public void init() {
+        deregisterVersionUsage(initialize());
+    }
+
+    private MVStore.TxCounter initialize() {
+        MVStore.TxCounter txCounter = mvStore.registerVersionUsage();
+        if (init.compareAndSet(false, true)) {
```

Comment:
there is a narrow race window here, at startup, where 2 JDBC connections both load LOB data.
The first one starts doing the map opening.
The second one bypasses that, and consequently might hit an NPE when accessing the lobMap/refMap fields.

Reply:
Yes, you are right, technically there is a race. I haven't changed that logic, just replaced plain boolean with atomic switch. But after a closer look, one might wonder - why that lazy logic is needed in the first place, since there is an explicit call to init() in Database(), right after LobStorageMap is constructed. It probably can / should go away.

Reply:
Initialization of LOB storage was delayed many years ago, but since 2deb0e7c548a44745a2eb6bc1630efe843f3c64d LOB storage is initialized during initialization of database.

Recently I refactored database initialization logic in 08ee2ec7a6af88a504c93a7ab546c10945211583, with my changes this situation became more visible.

It looks like `LobStorageInterface.init()` in both `LobStorageMap` and `LobStorageBackend` isn't really needed for years and implementations can be inlined into constructors.